File: route.py

from bottle import route, run, template, request, error, abort, default_app


TITLE = "Mikrotik DHCP option 249 route"

HTML_CODE = """
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <title>{{title or 'No title'}}</title>
</head>

<body>

<div class="jumbotron" style="padding: 0px 0px 0px 0px; margin-bottom: 10px;">
    <div class="container-fluid">
        <h1>Mikrotik</h1>
            <p>
                Make DHCP option 249 routes
            </p>
    </div>
</div>

<div class="bs-example">
    <form action="/mikrotik" method="POST">
        <textarea name="route" ROWS=20 COLS=70 placeholder="10.6.40.0/22-10.6.32.1,10.6.44.0/22-10.6.38.1"></textarea>
        </br>
        <INPUT TYPE=SUBMIT name="submit" VALUE="Submit">
    </form>
</div>

</body>
</html>
"""


@error(404)
def error404(error):
    return 'Error: %s' % error


@route('/', method='GET')
@route('/mikrotik', method='POST')
def index():
    if request.method == "GET":
        return template(HTML_CODE, title=TITLE)

    if request.method == "POST":
        try:
            form_data = request.forms.get('route')
        except Exception, e:
            abort(404, e)
        if form_data:
            route_dict = {}
            for i in form_data.split(","):
                subnet = i.split('-')[0]
                gateway = i.split('-')[1]
                route_dict[subnet] = gateway
            return routes2hex(route_dict)
        return "Empty data provided, please try again"


def ip2hex(cidr, router):
    addr, mask = cidr.split("/")
    mask = int(mask)
    addr_list = addr.split(".")
    hex_string = "%2s" % hex(int(mask))[2:].zfill(2)
    if mask > 0:
        hex_string += "%2s" % hex(int(addr_list[0]))[2:].zfill(2)
    if mask > 8:
        hex_string += "%2s" % hex(int(addr_list[1]))[2:].zfill(2)
    if mask > 16:
        hex_string += "%2s" % hex(int(addr_list[2]))[2:].zfill(2)
    if mask > 24:
        hex_string += "%2s" % hex(int(addr_list[3]))[2:].zfill(2)

    gw_split = router.split(".")
    hex_gw = "%2s%2s%2s%2s" % (
        hex(int(gw_split[0]))[2:].zfill(2),
        hex(int(gw_split[1]))[2:].zfill(2),
        hex(int(gw_split[2]))[2:].zfill(2),
        hex(int(gw_split[3]))[2:].zfill(2),
    )

    return "".join([hex_string, hex_gw])


def routes2hex(routes):
    routers = []
    for cidr, router in routes.items():
        a = ip2hex(cidr, router)
        routers.append(a)

    return "0x%s" % ("".join(routers).lower())


# Example
# print routes2hex({
#     "10.0.0.0/8": "10.6.34.1",
#     "172.17.1.0/24": "10.6.34.1"
# })

application = default_app()
# run(host='127.0.0.1', port=8090, debug=True, reloader=True)

File: route.pl
#!/usr/bin/perl -w

use strict;

    print "\033[2J";    #clear the screen
    print "\033[0;0H";  #jump to 0,0

sub option_121 {
    my $gw = shift;
    my $string = '';
    my ($subnet, $mask, $b0, $b1, $b2, $b3);
    foreach my $cidr (@_) {
        ($subnet,  $mask) = split('/', $cidr);
        ($b0, $b1, $b2, $b3) = split(/\./, $subnet);
        $string .= sprintf('%02x', $mask);
        $string .= sprintf('%02x', $b0) if($mask > 0);
        $string .= sprintf('%02x', $b1) if($mask > 8);
        $string .= sprintf('%02x', $b2) if($mask > 16);
        $string .= sprintf('%02x', $b3) if($mask > 24);
        $string .= sprintf('%02x%02x%02x%02x', split(/\./, $gw));
    }
    return $string;
}

if(@ARGV < 2)
{
    print "\n";
    print "Usage: $0 [gateway] [host|network]/[bitmask]\n\n";
    print "Example: $0 192.168.0.1 192.168.10.0/24\n\n";
}
elsif($ARGV[0] =~ /^\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$/)
{
    print "DHCP option 121 (249) hex string: ".option_121(@ARGV)."\n";
}
else
{
    print "Invalid gateway IP address: '$ARGV[0]'\n";
}
